### ISSUE TITLE
SQLite: Invalid SQL generated for multi-column primary keys

### ISSUE THREAD

**Issue by ondras** [ondras/wwwsqldesigner#97]
_From [NikitaTh...@gmail.com](https://code.google.com/u/112475953030429080148/) on March 24, 2010 04:46:07_

When I create a table with a multi-column (composite) primary key, WWW SQL
Designer generates SQL that SQLite won't accept. For example --

sqlite> CREATE TABLE foo (
   ...> field1 INTEGER NOT NULL PRIMARY KEY ,
   ...> field2 INTEGER NOT NULL PRIMARY KEY 
   ...> );
SQL error: table "foo" has more than one primary key

The correct syntax is this --

CREATE TABLE foo (
field1 INTEGER NOT NULL,
field2 INTEGER NOT NULL,
PRIMARY KEY (field1, field2)  
);

AFAICT that's the only way to express a composite primary key. 

Here's where it gets tricky.

In issue #65 I pointed out how AUTOINCREMENT can only be used following a
PRIMARY KEY constraint. In other words, the only permissible syntax for
AUTOINCREMENT primary keys differs from the syntax for composite primary keys:

CREATE TABLE foo (
id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT
);

For single-column, non-AUTOINCREMENT keys, either syntax is acceptable:
CREATE TABLE foo (
id INTEGER NOT NULL PRIMARY KEY 
);

CREATE TABLE foo (
id INTEGER NOT NULL,
PRIMARY KEY (id)
);

AUTOINCREMENT is not permitted on multi-column keys, so we can ignore that
case. (The SQLite docs say, "Multi-column primary keys do not become
aliases for the rowid. The AUTOINCREMENT keyword only works on a column
that is an alias for the rowid." http://www.sqlite.org/lang_createtable.html )

In short, it looks like one syntax is needed for AUTOINCREMENT primary keys
and another for composite primary keys. If someone attempts to create a
table that applies AUTOINCREMENT to one of the columns of a multi-column
primary key, it doesn't matter which syntax you use because SQLite won't
accept it.

**Attachment:** [invalid_example.xml](http://code.google.com/p/wwwsqldesigner/issues/detail?id=90)

_Original issue: http://code.google.com/p/wwwsqldesigner/issues/detail?id=90_

**Comment by ondras** [comment]
_From [ondrej.zara](https://code.google.com/u/ondrej.zara/) on March 24, 2010 03:39:40_

Hi,

thanks for reporting this as well as providing a very detailed explanation. If your
observations are correct, I will have to apply some major refactoring to the relevant
XSL file :/

**Status:** Accepted

**Comment by ondras** [comment]
_From [NikitaTh...@gmail.com](https://code.google.com/u/112475953030429080148/) on March 24, 2010 07:14:26_

Sorry to hear it.

There is one possible workaround, but I think if I saw the behavior I'm about to
describe in WWW SQL Designer, I'd file a bug on that too. 

SQLite's FAQ says, "A column declared INTEGER PRIMARY KEY will autoincrement". http://www.sqlite.org/faq.html#q1 I didn't know this, so I learned something new (and useful!) today.

HOWEVER, the behavior a column declared like so --
   id INTEGER PRIMARY KEY 
differs subtly from one declared like so --
   id INTEGER PRIMARY KEY AUTOINCREMENT

The latter definition guarantees that keys will be unique over the life of the table
whereas the first definition can reuse keys (although it still auto-generates unique
keys which is all that most people really want when they create an AUTOINCREMENT
column). 

Here's an example showing re-use of keys.

sqlite> CREATE TABLE foo (id INTEGER NOT NULL PRIMARY KEY, junk TEXT);
sqlite> INSERT INTO foo (junk) VALUES ('x');
sqlite> INSERT INTO foo (junk) VALUES ('y');
sqlite> INSERT INTO foo (junk) VALUES ('z');
sqlite> SELECT * FROM foo;
id          junk      
\----------  ----------
1           x         
2           y         
3           z         
sqlite> DELETE FROM foo WHERE id = 3;
sqlite> SELECT * FROM foo;
id          junk      
\----------  ----------
1           x         
2           y         
sqlite> INSERT INTO foo (junk) VALUES ('abc');
sqlite> SELECT * FROM foo;
id          junk      
\----------  ----------
1           x         
2           y         
3           abc       


Interestingly, Richard Hipp has pointed out that adding AUTOINCREMENT to a primary
key actually slows down SQLite a bit on inserts, because it has to check a table to
see what its next key will be.

So one could argue that AUTOINCREMENT is superfluous in SQLite and therefore WWW SQL
Designer doesn't need to include it in the SQL it generates for SQLite, but I don't
think that's a good workaround for this problem. IMO AUTOINCREMENT has three effects --
1) Slightly changes the behavior of the auto-generated keys.
2) Serves as documentation when looking at the CREATE TABLE statement.
3) Makes the generated SQL more portable to other databases.

In short, I think you're stuck with the ugly work of supporting the two different
syntaxes for PRIMARY KEY.

**Comment by ondras** [comment]
_From [ondrej.zara](https://code.google.com/u/ondrej.zara/) on March 24, 2010 07:38:49_

I fully agree. Omitting AUTOINCREMENT has very litle sense here; all your three
reasons are valid.

**Comment by ondras** [comment]
_From [ondrej.zara](https://code.google.com/u/ondrej.zara/) on April 08, 2010 23:46:46_

Fixed in r101 , please re-test. I also added support for SQLite foreign keys.

**Status:** Fixed  
  **Owner:** ondrej.zara

**Comment by ondras** [comment]
_From [NikitaTh...@gmail.com](https://code.google.com/u/112475953030429080148/) on April 12, 2010 06:11:54_

Works! The SQL that's generated is formatted a little oddly with the comma on its own
line (as in the example below), but it is syntactically correct and SQLite accepts
it. Thanks!

CREATE TABLE experiment_metabolites (
experiment_id TEXT NOT NULL  REFERENCES experiments (id)
,
metabolite_id TEXT NOT NULL  REFERENCES metabolites (id)
,
PRIMARY KEY (experiment_id, metabolite_id)
);

**Comment by ondras** [comment]
_From [ondrej.zara](https://code.google.com/u/ondrej.zara/) on April 12, 2010 06:35:11_

I see - a fix will be commited soon.

**Comment by itsho** [comment]
I'm not sure if this is the same issue or a new one.
but, when using this XML:

    <?xml version="1.0" encoding="utf-8" ?>
    
    
    <sql>
    <datatypes db="sqlite">
    	<group color="rgb(238,238,170)" label="Numeric">
    		<type label="Integer" quote="" sql="INTEGER" length="0"/>
    	 	<type label="TINYINT" quote="" sql="TINYINT" length="0"/>
    	 	<type label="SMALLINT" quote="" sql="SMALLINT" length="0"/>
    	 	<type label="MEDIUMINT" quote="" sql="MEDIUMINT" length="0"/>
    	 	<type label="INT" quote="" sql="INT" length="0"/>
    		<type label="BIGINT" quote="" sql="BIGINT" length="0"/>
    		<type label="Decimal" quote="" sql="DECIMAL" length="1" re="DEC"/>
    		<type label="Single precision" quote="" sql="FLOAT" length="0"/>
    		<type label="Double precision" quote="" sql="DOUBLE" length="0" re="DOUBLE"/>
    	</group>
    
    	<group color="rgb(255,200,200)" label="Character">
    		<type label="Char" quote="'" sql="CHAR" length="1"/>
    		<type label="Varchar" quote="'" sql="VARCHAR" length="1"/>
    		<type label="Text" quote="'" sql="MEDIUMTEXT" length="0" re="TEXT"/>
    		<type label="Binary" quote="'" sql="BINARY" length="1"/>
    		<type label="Varbinary" quote="'" sql="VARBINARY" length="1"/>
    		<type label="BLOB" quote="'" sql="BLOB" length="0" re="BLOB"/>
    	</group>
    
    	<group color="rgb(200,255,200)" label="Date &amp; Time">
    		<type label="Date" quote="'" sql="DATE" length="0"/>
    		<type label="Time" quote="'" sql="TIME" length="0"/>
    		<type label="Datetime" quote="'" sql="DATETIME" length="0"/>
    		<type label="Year" quote="" sql="YEAR" length="0"/>
    		<type label="Timestamp" quote="'" sql="TIMESTAMP" length="0"/>
    	</group>
    	<group color="rgb(200,200,255)" label="Miscellaneous">
    		<type label="ENUM" quote="" sql="ENUM" length="1"/>
    		<type label="SET" quote="" sql="SET" length="1"/>
    		<type label="Bit" quote="" sql="bit" length="0"/>
    	</group>
    </datatypes><table x="849" y="370" name="tracks">
    <row name="trackid" null="1" autoincrement="1">
    <datatype>INTEGER</datatype>
    <default>NULL</default></row>
    <row name="trackname" null="1" autoincrement="0">
    <datatype>MEDIUMTEXT</datatype>
    <default>NULL</default></row>
    <key type="PRIMARY" name="PRIM">
    <part>trackid</part>
    <part>trackname</part>
    </key>
    </table>
    </sql>


this is the resulted SQLite:

    CREATE TABLE 'tracks' (
        'trackid' INTEGER DEFAULT NULL AUTOINCREMENT,
        'trackname' MEDIUMTEXT DEFAULT NULL,
        PRIMARY KEY (trackid, trackname)
    );

when running this command I'm getting:

> Error while executing SQL query on database 'test': near "AUTOINCREMENT": syntax error

**Comment by ondras** [comment]
Hm, I am no SQLite guru. What is the correct syntax for this kind of table?

**Comment by itsho** [comment]
Seems like it's not supported, and the only solution is to use _UNIQUE_ **instead**:

`CREATE TABLE tracks (
    trackid   INTEGER    DEFAULT NULL PRIMARY KEY AUTOINCREMENT ,
    trackname MEDIUMTEXT DEFAULT NULL,
    UNIQUE (
        trackid,
        trackname
    )
)`

source:  http://stackoverflow.com/a/6157337/426315

**Comment by ondras** [comment]
Shall be fixed now. @itsho, can you please confirm?

**Comment by itsho** [comment]
yep. working well.

thanks!
